lib/triton_kernel_call.cc

```diff
@@ -0,0 +1,163 @@
+#include <pybind11/stl.h>
+#include "cuda.h"
+
+namespace py = pybind11;
+
+typedef std::map<std::string, py::object> asm_map_t;
+
+const int TRITON_MAX_N_SHARED_BYTES = 49152;
+const int TRITON_MAX_SHARED_OPTIN = 49152;
+
+
+class TritonCallDescriptor {
+  public:
+    explicit TritonCallDescriptor(std::string name, asm_map_t asm_map, std::uint32_t shared_mem,
+                                  std::uint32_t grid_0, std::uint32_t grid_1,
+                                  std::uint32_t grid_2, std::uint32_t num_warps,
+                                  std::uint32_t arity)
+      : name(std::move(name)),
+        asm_map(std::move(asm_map)),
+        shared_mem(shared_mem),
+        grid_0(grid_0),
+        grid_1(grid_1),
+        grid_2(grid_2),
+        num_warps(num_warps),
+        arity(arity),
+        kernels() {}
+    std::string name;
+    asm_map_t asm_map;
+    std::uint32_t shared_mem;
+    std::uint32_t grid_0;
+    std::uint32_t grid_1;
+    std::uint32_t grid_2;
+    std::uint32_t num_warps;
+    std::uint32_t arity;
+    std::map<CUdevice, CUfunction> kernels;
+
+    bool is_loaded(CUdevice device) {
+      return kernels.count(device) > 0;
+    }
+
+    void load(CUdevice device) {
+      // Mimics Triton kernel loading
+      std::string assembly;
+      if (asm_map.find("cubin") != asm_map.end())
+        assembly = py::cast<std::string>(asm_map["cubin"]);
+      else
+        assembly = py::cast<std::string>(asm_map["ptx"]);
+      CUfunction fun;
+      CUmodule mod;
+      cuModuleLoadData(&mod, assembly.c_str());
+      cuModuleGetFunction(&fun, mod, name.c_str());
+      int n_regs = 0;
+      int n_spills = 0;
+      cuFuncGetAttribute(&n_regs, CU_FUNC_ATTRIBUTE_NUM_REGS, fun);
+      cuFuncGetAttribute(&n_spills, CU_FUNC_ATTRIBUTE_LOCAL_SIZE_BYTES, fun);
+      n_spills /= 4;
+      int shared_optin;
+      cuDeviceGetAttribute(&shared_optin,
+                           CU_DEVICE_ATTRIBUTE_MAX_SHARED_MEMORY_PER_BLOCK_OPTIN,
+                           device);
+      if (shared_mem > TRITON_MAX_N_SHARED_BYTES &&
+          shared_optin > TRITON_MAX_SHARED_OPTIN) {
+        cuFuncSetCacheConfig(fun, CU_FUNC_CACHE_PREFER_SHARED);
+        int shared_total, shared_static;
+        cuDeviceGetAttribute(
+            &shared_total, CU_DEVICE_ATTRIBUTE_MAX_SHARED_MEMORY_PER_MULTIPROCESSOR,
+            device);
+        cuFuncGetAttribute(&shared_static, CU_FUNC_ATTRIBUTE_SHARED_SIZE_BYTES,
+                           fun);
+        cuFuncSetAttribute(fun, CU_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES,
+                           shared_optin - shared_static);
+      }
+      kernels[device] = fun;
+    };
+};
+
+void do_custom_call(CUstream stream, void** buffers,
+    char* opaque, size_t opaque_len) {
+  uint64_t descriptor = std::strtoull(opaque, NULL, 0);
+  TritonCallDescriptor* triton_call =
+      (TritonCallDescriptor*)(static_cast<uintptr_t>(descriptor));
+  int grid_0 = triton_call->grid_0;
```

Comment:
Just move all of this to be a function on the descriptor? Maybe rename TritionCallDescriptor to TritionExecutable?

Reply:
Done.